### PR TITLE
ConstExpr C++ API

### PR DESCRIPTION
**Pull request by arBmind**
Implements issue #50 

I created a third header. So you only use it if needed.

The following features are supported:
* Register Signals
* Implement Signals
* Register Properties

It might be extended if it's useful.

Limitations:
* No version of GCC is supported. It simply refuses to resolve the signals when used in properties.
* There is not much documentation beyond the unit test and inline documentation.

### CODE REVIEW

File: src/wobjectcpp.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
+#pragma once
+#include "wobjectdefs.h"
+
+namespace w_cpp {
```
Comment (ogoffart) [inline src/wobjectcpp.h]: `cpp` does not mean much in this respect.
Maybe `w_builder` would be a better name?

File: src/wobjectcpp.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+///
+/// example usage:
+///     template<size_t I>
+///     struct MyProperties;
+///     template<>
+///     struct MyProperties<0> {
+///         constexpr static auto property = w_cpp::makeProperty<QString>(w_cpp::viewLiteral("name"), w_cpp::viewLiteral("QString"))
+///             .setGetter(&Example::getName);
+///     };
+///     W_CPP_PROPERTY(MyProperties)
```
Comment (ogoffart) [inline src/wobjectcpp.h]: Would it be possible to have amacro such as this would be possible:

```
constexpr static auto propertyName = w_cpp::makeProperty<QString>(w_cpp::viewLiteral("name"), w_cpp::viewLiteral("QString"))
￼	///             .setGetter(&Example::getName);

W_ADD_PROPERTY(propertyName)
```

I don't really understand the meaning in CPP (because everything is CPP, so I suggest ADD instead.
And I think hafing to specialize a template is a bit complicated, if there are several properties, one can call W_ADD_PROPERTY several time


Perhaps even we should allow:

```
W_ADD_PROPERTY(w_cpp::makeProperty<QString>(w_cpp::viewLiteral("name"), w_cpp::viewLiteral("QString")))
```
￼	
For adding several property at once, a tuple could be used.

File: src/wobjectcpp.h [hunk 3]
```diff
@@ -0,0 +1,124 @@
+#pragma once
+#include "wobjectdefs.h"
+
+namespace w_cpp {
+
+using w_internal::StringView;
+using w_internal::viewLiteral;
+
+template<typename T>
+constexpr auto makeProperty(StringView name, StringView type) {
```
Comment (ogoffart) [inline src/wobjectcpp.h]: all the makeXXX needs documentation.  And also StringView

File: src/wobjectcpp.h [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    template<class... Args, class = std::enable_if_t<(std::is_same_v<std::decay_t<Args>, StringView> && ...)>>
+#else
+    template<class... Args, class = std::enable_if_t<all<std::is_same<std::decay_t<Args>, StringView>::value...>()>>
+#endif
+    constexpr auto setParamNames(Args... paramNames) const
+        -> MetaMethodInfoBuilder<F, Flags, IC, ParamTypes, w_internal::StringViewArray<sizeof... (Args)>> {
+        return {name, func, paramTypes, {paramNames...}};
+    }
+    template<int... fs>
+    constexpr auto addFlags(w_internal::W_MethodFlags<fs>...) const
```
Comment (ogoffart) [inline src/wobjectcpp.h]: Is the user supposed to reach into the internal namespace?

File: src/wobjectcpp.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    friend constexpr auto w_state(w_internal::Index<I>, w_internal::SignalStateTag, W_ThisType**) W_RETURN((a<I-W_MACRO_CONCAT(a,_O)>::signal))
+
+/// \macro W_CPP_SIGNAL_IMPL(type, callback, index)
+/// allows to implement a signal
+///
+/// example usage:
+///     template<size_t I>
+///     void notifyPropertyChanged() W_CPP_SIGNAL_IMPL(decltype (&tst_CppApi::notifyPropertyChanged<I>), MySignals, I, 0)
+///
+#define W_CPP_SIGNAL_IMPL(type, a, i, ...) { \
```
Comment (ogoffart) [inline src/wobjectcpp.h]: I am not really understanding the use case.

This does not really works if the signal has different type.

Perhaps the best way to emit the signal is to simply call QMetaObject::activate somehow

**Comment by arBmind** [comment]
@ogoffart Thank you for your review. It seems I have not done a very good job of transporting my reasoning.

If a user has only a single property, he should be able to use the established macro based API.
The motivation for the C++ constexpr based API is to allow us to add multiple properties without too much boilerplate.

File: src/wobjectcpp.h (hunk 1 above)
Comment (arBmind) [reply src/wobjectcpp.h]: I'm not a fan of it either - just needed a name.
Maybe 
* `w_cpp` or `w_constexpr` to constrast it as C++ or Constexpr based API.
* `w_object` - to show the relation to W_OBJECT

File: src/wobjectcpp.h [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+using w_internal::StringView;
+using w_internal::viewLiteral;
+
+template<typename T>
+constexpr auto makeProperty(StringView name, StringView type) {
+    return w_internal::MetaPropertyInfo<T>{ name, type, {}, {}, {}, {}, {} };
+}
+
+#if __cplusplus <= 201700L
+template<bool... Bs>
```
[the remark this answers is not in the document]
Comment (arBmind) [reply src/wobjectcpp.h]: agreed… moved to `w_internal`

File: src/wobjectcpp.h (hunk 3 above)
Comment (arBmind) [reply src/wobjectcpp.h]: done

File: src/wobjectcpp.h (hunk 4 above)
Comment (arBmind) [reply src/wobjectcpp.h]: That is not necessary.

[wobjectdefs.h#L185](https://github.com/woboq/verdigris/blob/master/src/wobjectdefs.h#L185) and follows, those flags are given public names like `W_Access::Private`

But the types are still `w_internal::W_MethodFlags`.

File: src/wobjectcpp.h (hunk 5 above)
Comment (arBmind) [reply src/wobjectcpp.h]: The `W_SIGNAL` macro generates both an implementation of the signal function and the registration.

The registration is easily done with the constexpr hook. But the implementation part would be missing.
This macro allows the generation of the signal function body. Inspired by [wobjectdefs.h#L788](https://github.com/woboq/verdigris/blob/master/src/wobjectdefs.h#L788).

It's possible that there is a bug I don't see. But it works well for our use case of property changed signals without any arguments.

File: src/wobjectcpp.h (hunk 2 above)
Comment (arBmind) [reply src/wobjectcpp.h]: `W_CPP_` is just the prefix I used. It's basically the C++ way to add properties.

Specializing a template makes this approach very accessible. Users of this API most likely will have their own constexpr based way to represent the properties. We don't know what they use. So they have to provide an adapter.

Internally we count everything. So this approach is very light weight and adds almost no compile time.

If we want to allow a tuple based approach, we would force the compiler to create a complex constexpr enable tuple, just to decompose it into something that is indexed. This would make the API more complicated, harder to use and slower to compile.

**Comment by ogoffart** [comment]
I still don't see the problem with a tuple as it is easy to build and desconstruct from constexpr code.

It would probably help me to see the concreate usecase you have for this to help me understanding.

**Comment by arBmind** [comment]
Ok, I will try to make my use case a bit more obvious.

We use the Type Driven Approach to write a distributed software (see presentation https://github.com/arBmind/2019-types-de) and want to generate QML enabled models.

TLDR: We have a strong type for all properties and want to make them accessible via QObject properties.

For each property passed to TupleObject we store the `Output` type and constexpr value `name`.

The code looks basically like this.
```cpp
template<class... Ts>
class TupleObject : public QObject {
    W_OBJECT(TupleObject)

public:
    using PropertyPack = TypePack<typename Ts::Output ...>; // + some extra Properties

    template<size_t I>
    constexpr static auto propertyName() {
        using T = UnwrapType< TypeAt< Index<I>, TypePack<Ts...> >>;
        return T::name;
    }

    template<class T>
    auto propertyValue() const;

    template<size_t I>
    void notifyPropertyChanged() W_CPP_SIGNAL_IMPL(decltype(&TupleObject::notifyPropertyChanged<I>), Signals, I, 0)

    template<size_t I, class = std::enable_if_t<(I < meta17::count_types<PropertyPack>)>>
    struct Signals {
        constexpr static auto signalName = meta_verdigris::stringAddNull(
	    string17::concat(TupleObject::propertyName<I>(), string17::storeLiteral("Changed")));
        constexpr static auto signal =
            w_cpp::makeSignalBuilder(toVerdigrisString(signalName), &TupleObject::notifyPropertyChanged<I>).build();
    };
    W_CPP_SIGNAL(Signals)

    template<size_t I, class = std::enable_if_t<(I < meta17::count_types<PropertyPack>)>>
    struct Properties {
        using T = UnwrapType< TypeAt< Index<I>, PropertyPack >>;
        using PropertyType = decltype(std::declval<TupleObject>().template propertyValue<T>());

        constexpr static auto propertyName = meta_verdigris::stringAddNull(TupleObject::propertyName<I>());

        constexpr static auto property = [] {
            return w_cpp::makeProperty<PropertyType>(
                       toVerdigrisString(propertyName), w_cpp::viewLiteral("Custom"))
                .setGetter(&TupleObject::propertyValue<T>)
                .setNotify(&TupleObject::notifyPropertyChanged<I>);
        }();
    };
    W_CPP_PROPERTY(Properties)
};
```

If we would accept the Properties with a tuple-like interface. The user code will have to run a loop through all the properties. The verdigris side will have to implement a way to store each value as `w_state`. So both sides would have to do more work. And more work also compiles slower.

Also keep in mind that `std::tuple` is not constexpr enabled. We would have to implement a constexpr enabled version just for this use case.

It's doable no question, but from my experience it wont be any better. Just easier to document, but harder to use and implement.

File: src/wobjectcpp.h [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+///     constexpr char text[6] = "Hello"; // if you have a string literal use `viewLiteral` below.
+///     constexpr auto view = w_cpp::StringView{&text[0], &text[6]};
+///
+/// \note the end pointer has to point behind the \0
+using w_internal::StringView;
+
+/// Generate a constexpr StringView from a C++ string literal.
+///
+/// example usage:
+///     constexpr auto view & w_cpp::StringView{"Hello"};
```
Comment (ogoffart) [inline src/wobjectcpp.h]: did you mean this:

```suggestion
///     constexpr auto view = w_cpp::viewLiteral{"Hello"};
```

Should class name not start with upper-case?

File: src/wobjectcpp.h [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+using w_internal::viewLiteral;
+
+/// create a compile time property description for registraton usage
+///
+/// \arg T the type of the property
+/// \arg name the compile time StringView for the property name
+/// \arg type the compile time StringView for the type of the property
+///
+/// Following methods may be called in chain:
```
Comment (ogoffart) [inline src/wobjectcpp.h]: maybe mention it is using the "Builder pattern"

File: tests/cppapi/tst_cppapi.cpp [hunk 9]
```diff
@@ -0,0 +1,140 @@
+#include <wobjectcpp.h>
+#include <wobjectimpl.h>
+#include <QtCore/QMetaObject>
+#include <QtTest/QtTest>
+
+class tst_CppApi : public QObject
+{
+    W_OBJECT(tst_CppApi)
+
+#if !defined(Q_CC_GNU) || defined(Q_CC_CLANG) // GCC refuses to properly resolve registered signals for properties
```
Comment (ogoffart) [inline tests/cppapi/tst_cppapi.cpp]: So this does not work with GCC?
That's quite bad.

Comment (arBmind) [reply tests/cppapi/tst_cppapi.cpp]: Yes, unfortunately, I found no way to convince GCC to work with this.

For my project we do not really care about GCC - so it may be that I am missing some tricks there.

File: src/wobjectcpp.h (hunk 1 above)
Comment (arBmind) [reply src/wobjectcpp.h]: @ogoffart We still need to decide on the namespace name.
Do you have a preferred name?

File: src/wobjectcpp.h (hunk 7 above)
Comment (arBmind) [reply src/wobjectcpp.h]: Thank you for pointing this out. A stupid copy & paste error.

viewLiteral is actually a function. Therefore it starts with a lower case letter.

File: src/wobjectcpp.h (hunk 8 above)
Comment (arBmind) [reply src/wobjectcpp.h]: Its not really a builder pattern, as no build function has to be called. Each step is valid.
If you want I can implement a real builder for properties, just for this new API.
For the first implementation it seemed a bit too much work.

@ogoffart What do you think?

**Comment by arBmind** [comment]
Ok, I added a tutorial. For now it's only compiled with Qbs. I forgot how to write QML applications with QMake.

File: src/wobjectcpp.h (hunk 8 above)
Comment (ogoffart) [reply src/wobjectcpp.h]: I think that's fine as it is.

File: tests/cppapi/tst_cppapi.cpp (hunk 9 above)
Comment (ogoffart) [reply tests/cppapi/tst_cppapi.cpp]: GCC is one of the most important compiler. What exactly is the problem?

File: src/wobjectcpp.h (hunk 1 above)
Comment (ogoffart) [reply src/wobjectcpp.h]: `w_object` is not bad.  since the macro is uppercase, it does not clash.
Since this allow to create properties, `w_builder` or `w_propertybuilder`
Or maybe `w_meta` ?

File: tests/cppapi/tst_cppapi.cpp (hunk 9 above)
Comment (arBmind) [reply tests/cppapi/tst_cppapi.cpp]: As I wrote in the limitations section of the PR:
> * No version of GCC is supported. It simply refuses to resolve the signals when used in properties.

I tested it again… the new tutorial seems to work fine with GCC 7.3. But the test does still not compile.
